// coc-omnisharp/out/client/extension.js
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const coc_nvim_1 = require("coc.nvim");
const coc_nvim_2 = require("coc.nvim");
const coc_utils_1 = require("coc-utils");
const logger = coc_nvim_1.workspace.createOutputChannel("coc-omnisharp");
const omnisharpRepo = {
    kind: "github",
    repo: "omnisharp/omnisharp-roslyn",
    channel: "latest"
};
const omnisharpPacks = {
    "win-x64": { platformPath: "omnisharp-win-x64.zip", executable: "Omnisharp.exe" },
    "linux-x64": { platformPath: "omnisharp-linux-x64.zip", executable: "run" },
    "osx-x64": { platformPath: "omnisharp-osx.zip", executable: "run" },
};
function activate(context) {
    return __awaiter(this, void 0, void 0, function* () {
        logger.appendLine("coc-omnisharp activated.");
        logger.appendLine(`workspace root=${coc_nvim_1.workspace.rootPath}`);
        // Options to control the language client
        let clientOptions = {
            // Register the server for C#/VB documents
            documentSelector: [{ scheme: 'file', language: 'cs' }, { scheme: 'file', language: 'vb' }],
            synchronize: {
                configurationSection: 'omnisharp',
                fileEvents: [
                    coc_nvim_1.workspace.createFileSystemWatcher('**/*.cs'),
                    coc_nvim_1.workspace.createFileSystemWatcher('**/*.csx'),
                    coc_nvim_1.workspace.createFileSystemWatcher('**/*.cake'),
                    coc_nvim_1.workspace.createFileSystemWatcher('**/*.vb')
                ]
            }
        };
        const omnisharpProvider = new coc_utils_1.LanguageServerProvider(context, "OmniSharp", omnisharpPacks, omnisharpRepo);
        const omnisharpExe = yield omnisharpProvider.getLanguageServer();
        const config = coc_nvim_1.workspace.getConfiguration('omnisharp');
        let serverOptions = {
            command: omnisharpExe,
            args: ["-lsp"],
            options: { cwd: coc_nvim_1.workspace.rootPath }
        };
        // Create the language client and start the client.
        let client = new coc_nvim_2.LanguageClient('cs', 'OmniSharp Language Server', serverOptions, clientOptions);
        let client_dispose = client.start();
        // Push the disposable to the context's subscriptions so that the 
        // client can be deactivated on extension deactivation
        context.subscriptions.push(client_dispose);
        let cmd_updateomnisharp = coc_nvim_1.commands.registerCommand('omnisharp.downloadLanguageServer', () => __awaiter(this, void 0, void 0, function* () {
            if (client.started) {
                yield client.stop();
                client_dispose.dispose();
                yield coc_utils_1.sleep(1000);
            }
            yield omnisharpProvider.downloadLanguageServer();
            client_dispose = client.start();
            context.subscriptions.push(client_dispose);
        }));
        context.subscriptions.push(cmd_updateomnisharp);
    });
}
exports.activate = activate;
